src/rdkafka_msgset_reader.c

```diff
@@ -983,24 +979,62 @@ rd_kafka_msgset_reader (rd_kafka_msgset_reader_t *msetr) {
                                 msetr->msetr_tver->version, rktp, Offset,
                                 "Unsupported Message(Set) MagicByte %d "
                                 "at offset %"PRId64,
-                                (int)MagicByte, Offset);
+                                (int)*MagicBytep, Offset);
                         /* Skip message(set) */
                         msetr->msetr_rktp->rktp_offsets.fetch_offset = Offset+1;
                 }
 
                 return RD_KAFKA_RESP_ERR__NOT_IMPLEMENTED;
         }
 
-        /* Let version-specific reader parse MessageSets until the slice
-         * is exhausted or an error occurs (typically a partial message). */
+        return RD_KAFKA_RESP_ERR_NO_ERROR;
+
+ err_parse:
+        return RD_KAFKA_RESP_ERR__BAD_MSG;
+}
+
+
+/**
+ * @brief Parse and read messages from msgset reader buffer.
+ */
+static rd_kafka_resp_err_t
+rd_kafka_msgset_reader (rd_kafka_msgset_reader_t *msetr) {
+        rd_kafka_buf_t *rkbuf = msetr->msetr_rkbuf;
+        rd_kafka_resp_err_t (*reader[])
+                (rd_kafka_msgset_reader_t *) = {
+                /* Indexed by MsgVersion/MagicByte, pointing to
+                 * a Msg(Set)Version reader */
+                [0] = rd_kafka_msgset_reader_msg_v0_1,
+                [1] = rd_kafka_msgset_reader_msg_v0_1,
+                [2] = rd_kafka_msgset_reader_v2
+        };
+        rd_kafka_resp_err_t err;
+
+        /* Parse MessageSets until the slice is exhausted or an
+         * error occurs (typically a partial message). */
         do {
+                int8_t MagicByte;
+
+                /* We dont know the MsgVersion at this point, peek where the
+                 * MagicByte resides both in MsgVersion v0..1 and v2 to
+                 * know which MessageSet reader to use. */
+                err = rd_kafka_msgset_reader_peek_msg_version(msetr,
+                                                              &MagicByte);
+                if (unlikely(err)) {
+                        if (err == RD_KAFKA_RESP_ERR__BAD_MSG)
+                                /* Read underflow, not an error. */
```

Comment:
Presumably the underflow is not considered an error because of your comment in #2090?

>Underflow (e.g., short responses) are expected due to the way the broker uses sendfile.

If so clarifying this in the comment would be nice.